### ISSUE TITLE
False negative type error for async middleware

### ISSUE THREAD

**Issue by markedwards** [blomqma/next-rest-framework#163]
There seems to be a subtle problem with the typing of `RouteMiddleware` if the function is async, and it can return both a `NextResponse` and some custom options.

Example:
```typescript
import { TypedNextResponse, route, routeOperation } from "next-rest-framework";

async function sleep() {
  await new Promise(() => setTimeout(() => {}, 1000));
}

export const { GET } = route({
  hello: routeOperation({
    method: "GET",
  })
    .middleware(async () => {
      await sleep();
      if (Math.random() > 0.5) {
        return {
          foo: "bar",
        };
      }
      return TypedNextResponse.json(
        {
          message: "",
        },
        {
          status: 400,
        },
      );
    })
    .handler((_req, _ctx, { foo }) => {
      console.log(foo);
    }),
});
```

This raises `Type '{ foo: string; }' is not assignable to type 'void | NextResponse<unknown> | TypedNextResponseType<unknown, number, AnyContentTypeWithAutocompleteForMostCommonOnes>'. ts(2345)`.  Also, the `foo` argument in the handler is `unknown`.

If the function is made sync (and the `await sleep()` line is removed, there's no type error.  So it seems to be some unimplemented type condition in the async case.

**Comment by blomqma** [comment]
Thanks, this should be fixed in `v6.0.0`.

**Comment by markedwards** [comment]
Confirmed this is fixed.  One slightly odd thing is I have to wrap my object like so:
```typescript
      return {
        ...obj,
      };
```

Otherwise I get `Type 'Foo' is not assignable to type 'BaseOptions'. Index signature for type 'string' is missing in type 'Foo'`.  Presumably because the object is just described by an interface and does not inherit from `Record<string, unknown>`?

Otherwise it works great though.  Thanks!
